drivers/usb/udc/udc_mcux_ehci.c

```diff
@@ -0,0 +1,927 @@
+	}
+
+	k_mutex_init(&data->mutex);
+
+	for (int i = 0; i < config->num_of_eps; i++) {
+		config->ep_cfg_out[i].caps.out = 1;
+		if (i == 0) {
+			config->ep_cfg_out[i].caps.control = 1;
+			config->ep_cfg_out[i].caps.mps = 64;
+		} else {
+			config->ep_cfg_out[i].caps.bulk = 1;
+			config->ep_cfg_out[i].caps.interrupt = 1;
+			config->ep_cfg_out[i].caps.iso = 1;
+			config->ep_cfg_out[i].caps.mps = 1024;
+		}
+
+		config->ep_cfg_out[i].addr = USB_EP_DIR_OUT | i;
+		err = udc_register_ep(dev, &config->ep_cfg_out[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	for (int i = 0; i < config->num_of_eps; i++) {
+		config->ep_cfg_in[i].caps.in = 1;
+		if (i == 0) {
+			config->ep_cfg_in[i].caps.control = 1;
+			config->ep_cfg_in[i].caps.mps = 64;
+		} else {
+			config->ep_cfg_in[i].caps.bulk = 1;
+			config->ep_cfg_in[i].caps.interrupt = 1;
+			config->ep_cfg_in[i].caps.iso = 1;
+			config->ep_cfg_in[i].caps.mps = 1024;
+		}
+
+		config->ep_cfg_in[i].addr = USB_EP_DIR_IN | i;
+		err = udc_register_ep(dev, &config->ep_cfg_in[i]);
+		if (err != 0) {
+			LOG_ERR("Failed to register endpoint");
+			return err;
+		}
+	}
+
+	data->caps.rwup = true;
+	data->caps.mps0 = USB_MCUX_MPS0;
+	data->caps.hs = true;
+	priv->dev = dev;
+
+	pinctrl_apply_state(config->pincfg, PINCTRL_STATE_DEFAULT);
+
+	return 0;
+}
+
+static const struct udc_api udc_mcux_api = {
+	.device_speed = udc_mcux_device_speed,
+	.ep_enqueue = udc_mcux_ep_enqueue,
+	.ep_dequeue = udc_mcux_ep_dequeue,
+	.ep_set_halt = udc_mcux_ep_set_halt,
+	.ep_clear_halt = udc_mcux_ep_clear_halt,
+	.ep_try_config = NULL,
+	.ep_enable = udc_mcux_ep_enable,
+	.ep_disable = udc_mcux_ep_disable,
+	.host_wakeup = udc_mcux_host_wakeup,
+	.set_address = udc_mcux_set_address,
+	.enable = udc_mcux_enable,
+	.disable = udc_mcux_disable,
+	.init = udc_mcux_init,
+	.shutdown = udc_mcux_shutdown,
+	.lock = udc_mcux_lock,
+	.unlock = udc_mcux_unlock,
+};
+
+/* EHCI device driver interface */
+static const usb_device_controller_interface_struct_t udc_mcux_if = {
+	USB_DeviceEhciInit, USB_DeviceEhciDeinit, USB_DeviceEhciSend,
+	USB_DeviceEhciRecv, USB_DeviceEhciCancel, USB_DeviceEhciControl
+};
+
+#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED
+#define UDC_MCUX_PHY_DEFINE(n)								\
+static usb_phy_config_struct_t phy_config_##n = {					\
+	.D_CAL = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx_d_cal, 0),		\
+	.TXCAL45DP = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx_cal_45_dp_ohms, 0),	\
+	.TXCAL45DM = DT_PROP_OR(DT_INST_PHANDLE(n, phy_handle), tx_cal_45_dm_ohms, 0),	\
+}
+
+#define UDC_MCUX_PHY_DEFINE_OR(n)							\
+	COND_CODE_1(DT_NODE_HAS_PROP(DT_DRV_INST(n), phy_handle),			\
+		    (UDC_MCUX_PHY_DEFINE(n)), ())
+
+#define UDC_MCUX_PHY_CFG_PTR_OR_NULL(n)							\
+	.phy_config = COND_CODE_1(DT_NODE_HAS_PROP(DT_DRV_INST(n), phy_handle),		\
+		    (&phy_config_##n), (NULL))
+#else
+#define UDC_MCUX_PHY_DEFINE_OR(n)
+#define UDC_MCUX_PHY_CFG_PTR_OR_NULL(n)
+#endif
```

Comment:
Not addressed, `#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED` is not required.

Reply:
I don't understand. In one platform that doesn't have the PHY driver, so the usb_phy.h is not present/valid in this platform. Why should I include this file?
I use `CONFIG_DT_HAS_NXP_USBPHY_ENABLED` to reduce the `phy_config` here because it is not needed if the platform doesn't have this phy.

Reply:
I can remove all the phy related codes from the driver and put it in the soc.c of different platforms. The previous version NXP usb device controller driver (usb_dc_mcux.c) works this way. 
Please let me know your thought.

Reply:
The phy initialization code is fine and belongs to the shim driver. There is nothing in [phy/usb_phy.h](https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/middleware/mcux-sdk-middleware-usb/phy/usb_phy.h) that requires you to include it conditionally, it can always build. Therefore, all your `#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED` can be removed, resulting in much cleaner code without unnecessary #ifdef. Here it is doubly unnecessary, whatever phy is present and needs to be initialized is covered by the macros above.

Reply:
Hi Johann, In SDK USB stack, I cannot guarantee the usb_phy.h will not contain the soc/platform/IP level codes in future even it only contains the common C codes now. The reason is: NXP releases SDK package by platform, one platform has one package. usb_phy.h is only present in the package of platforms that have this PHY, and it is only valid/used in the platform, it is never built with the platforms that doesn't have the PHY. Without NXP SDK release's pre-building, I cannot guarantee usb_phy.h will build successfully with the platforms that doesn't have this phy forever.

Reply:
@MarkWangChinese Got it, can you please add a comment before `#include "usb_phy.h`" that the header is not available for every package, thanks.

But, `#ifdef CONFIG_DT_HAS_NXP_USBPHY_ENABLED ... #else` is not necessary here, see https://github.com/jfischer-no/zephyr/commit/7652d1647de7dbf1cdce4b6f706605f6ea8c9e48 (https://github.com/jfischer-no/zephyr/commit/7652d1647de7dbf1cdce4b6f706605f6ea8c9e48.patch)

Reply:
Dev Review discussion:
@MarkWangChinese we discussed this PR with @jfischer-no in today's dev review meeting. He is would like to see the #ifdef removed and will approve the PR for merging. If we find later on subsequent platforms this is a problem we can address then.

Johann also re-kicked the CI so that is clean. So we will be able to merge this PR on Friday 6/14.

Reply:
@dleach02 @MarkWangChinese Actually let get it in, we have CI green and do not know what happens tomorrow :smiley:, https://github.com/zephyrproject-rtos/zephyr/pull/70401#discussion_r1638374321 can be follow up.

Reply:
I have created one PR https://github.com/zephyrproject-rtos/zephyr/pull/74382, @jfischer-no please help to review.